services/submission/src/main/java/app/coronawarn/server/services/submission/checkins/EventCheckinDataFilter.java

```diff
@@ -37,10 +43,24 @@ public EventCheckinDataFilter(SubmissionServiceConfig submissionServiceConfig,
    * <li>Filter out checkins which have trace location signatures that can not be verified</li>.
    */
   public List<CheckIn> filter(List<CheckIn> checkins) {
-    return checkins.stream()
+    AtomicInteger checkinsPickedAfterTrlFiltering = new AtomicInteger();
+    AtomicInteger checkinsPickedAfterOldFiltering = new AtomicInteger();
+    AtomicInteger checkinsPickedAfterFutureFiltering = new AtomicInteger();
+    var filtered = checkins.stream()
         .filter(this::filterOutZeroTransmissionRiskLevel)
+        .peek(k -> checkinsPickedAfterTrlFiltering.incrementAndGet())
         .filter(this::filterOutOldCheckins)
-        .filter(this::filterOutFutureCheckins).collect(Collectors.toList());
+        .peek(k -> checkinsPickedAfterOldFiltering.incrementAndGet())
+        .filter(this::filterOutFutureCheckins)
+        .peek(k -> checkinsPickedAfterFutureFiltering.incrementAndGet())
```

Comment:
Peek is not intended for code with side-effects and this will most likely break in Java 9 (see [here](https://stackoverflow.com/a/48221925)). Although it doesn't look as nice as your full-on stream solution, you should probably consider something like this:

```Java
int checkinsBeforeFiltering = checkins.size();

var List<CheckIn> filtered = checkins.stream().filter(this::filterOutZeroTransmissionRiskLevel).collect(Collectors.toList());
int checkinsPickedAfterTrlFiltering = filtered.size();

filtered = checkins.stream().filter(this::filterOutOldCheckins).collect(Collectors.toList());
int checkinsPickedAfterOldFiltering = filtered.size();

filtered = checkins.stream().filter(this::filterOutFutureCheckins).collect(Collectors.toList());
int checkinsPickedAfterFutureFiltering = filtered.size();
```

Reply:
On the other hand, the same is already being done in the [FederationUploadKeyService](https://github.com/corona-warn-app/cwa-server/blob/b75ba8e12bff38c878f4ada328c1cf8b08559281/common/persistence/src/main/java/app/coronawarn/server/common/persistence/service/FederationUploadKeyService.java#L60). So, let's not start a refactoring now.

Reply:
Good point though. Thanks for raising awareness. It kind of falls in the technical debt backlog for now.